File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, p: List[int]) -> int:
        # 7 , 1 , 5 , 3 , 6 , 4
        # whenever u come across a better deal, 
        # that is when the price u brought at x day is higher than the current price , we buy in the current day , and look for higher prices in the future days
        profit = 0
        m = p[0]
        for i in range(1 , len(p)):
            if m < p[i] :
                profit = max ( profit , p[i] - m)
            else:
                m = p[i]
        return profit

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int mini=prices[0];
        int profit=0;
        for(int i=1;i<prices.length;i++){
            if(prices[i]>mini){
                profit=Math.max(profit,prices[i]-mini);
            }
            else{
                mini=prices[i];
            }
        }
        return profit;
    }
}